pkg/ec2provider/ec2provider_test.go

```diff
@@ -0,0 +1,148 @@
+package ec2provider
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+)
+
+const (
+	DescribeDelay = 100
+)
+
+type mockEc2Client struct {
+	ec2iface.EC2API
+	Reservations []*ec2.Reservation
+}
+
+func (c *mockEc2Client) DescribeInstances(in *ec2.DescribeInstancesInput) (*ec2.DescribeInstancesOutput, error) {
+	// simulate the time it takes for aws to return
+	time.Sleep(DescribeDelay * time.Millisecond)
+	var reservations []*ec2.Reservation
+	for _, res := range c.Reservations {
+		var reservation ec2.Reservation
+		for _, inst := range res.Instances {
+			for _, id := range in.InstanceIds {
+				if aws.StringValue(id) == aws.StringValue(inst.InstanceId) {
+					reservation.Instances = append(reservation.Instances, inst)
+				}
+			}
+		}
+		if len(reservation.Instances) > 0 {
+			reservations = append(reservations, &reservation)
+		}
+	}
+	return &ec2.DescribeInstancesOutput{
+		Reservations: reservations,
+	}, nil
+}
+
+func setup() *ec2ProviderImpl {
+	dnsCache := ec2PrivateDNSCache{
+		dnsCache:     make(map[string]string),
+		dnsCacheLock: sync.RWMutex{},
+	}
+	ec2RequestQueue := ec2RequestQueue{
+		requestQueueMap:  make(map[string]bool),
+		requestQueueLock: sync.RWMutex{},
+	}
+	return &ec2ProviderImpl{
+		ec2:                &mockEc2Client{},
+		privateDNSCache:    dnsCache,
+		requestQueue:       ec2RequestQueue,
+		instanceIdsChannel: make(chan string, max_channel_size),
+	}
+
+}
+
+func TestGetPrivateDNSName(t *testing.T) {
+	ec2Provider := setup()
+	ec2Provider.ec2 = &mockEc2Client{Reservations: prepareSingleInstanceOutput()}
+	go ec2Provider.StartEc2DescribeBatchProcessing()
+	dns_name, err := ec2Provider.GetPrivateDNSName("ec2-1")
+	if err != nil {
+		t.Error("There is an error which is not expected when calling ec2 API with setting up mocks")
+	}
+	if dns_name != "ec2-dns-1" {
+		t.Errorf("want: %v, got: %v", "ec2-dns-1", dns_name)
+	}
+}
+
+func prepareSingleInstanceOutput() []*ec2.Reservation {
+	var reservations []*ec2.Reservation
+	instance := &ec2.Instance{
+		InstanceId:     aws.String("ec2-1"),
+		PrivateDnsName: aws.String("ec2-dns-1"),
+	}
+	var instances []*ec2.Instance
+	instances = append(instances, instance)
+	res1 := &ec2.Reservation{
```

Comment:
instead of all the variables and appends, what do you think about just returning a single variable with the nested fields set?

Reply:
good idea let me check.